### PR TITLE
Make Transformer tolerate missing layers for PP

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #318
* __->__ #322
* #321

A few small changes here lets manual PP frontend 'reconfigure' a whole
transformer model to a stage's portion simply by setting undesired
layers to None (in cases of top level layers) or deleting them from the
ModuleDict (for 'layers.*').

These changes don't impact the FQNs of the remaining layers, which is
critical for checkpoint load/save compatibility.

### CODE REVIEW

File: torchtitan/models/llama/model.py [hunk 1]
```diff
@@ -425,12 +425,14 @@ def forward(self, tokens: torch.Tensor):
             torch.Tensor: Output logits after applying the Transformer model.
 
         """
-        h = self.tok_embeddings(tokens)
+        # passthrough for nonexistent layers, allows easy configuration of pipeline parallel stages
+        h = self.tok_embeddings(tokens) if self.tok_embeddings else tokens
 
-        for layer in self.layers:
+        for layer in self.layers.values():
```
Comment (tianyu-l) [inline torchtitan/models/llama/model.py:431]: Is order still respected after switching to dict? If not, we need to sort the layers based on `int(key)`.

Comment (wconstab) [reply torchtitan/models/llama/model.py:431]: yea, it is. https://pytorch.org/docs/stable/generated/torch.nn.ModuleDict.html

**Comment by fegin** [comment]
Nice. But it is less intuitive than I originally thought. Especially the int/str conversion part. Not sure if that's a best UX for pippy or a customized PipelineModuleList will be easier for users.

**Review by fegin — APPROVED** [review]

**Review by wanchaol — APPROVED** [review]
lgtm!

File: torchtitan/models/llama/model.py [hunk 2]
```diff
@@ -366,9 +366,9 @@ def __init__(self, model_args: ModelArgs):
         # just the non-persistent buffers that is called after loading checkpoints.
         self.register_buffer("freqs_cis", self._precompute_freqs_cis(), persistent=True)
 
-        self.layers = torch.nn.ModuleList()
+        self.layers = torch.nn.ModuleDict()
         for layer_id in range(model_args.n_layers):
-            self.layers.append(TransformerBlock(layer_id, model_args))
+            self.layers[str(layer_id)] = TransformerBlock(layer_id, model_args)
```
Comment (awgu) [inline torchtitan/models/llama/model.py:371]: curious why do the dict keys have to be `str` (as opposed to `int` directly)?

**Comment by awgu** [comment]
One downside to using `ModuleDict` is that now the model print does not collapse `TransformerBlock`s together, making the model print very long.

**Pull request merged**